src/ssa/dynamic_objects.cpp

```diff
@@ -0,0 +1,568 @@
+/*******************************************************************\
+
+Module: Dynamically allocated objects
+
+Author: Viktor Malik
+
+\*******************************************************************/
+
+#include "dynamic_objects.h"
+#include "dynobj_instance_analysis.h"
+#include "ssa_value_set.h"
+
+#include <analyses/constant_propagator.h>
+#include <util/c_types.h>
+#include <util/expr_util.h>
+#include <util/type.h>
+#include <util/pointer_offset_size.h>
+#include <util/arith_tools.h>
+
+#include <algorithm>
+#include <iostream>
+
+dynamic_objectt::dynamic_objectt(
+  const goto_programt::instructiont *loc,
+  const typet &type,
+  const std::string &suffix,
+  bool concrete)
+  :loc(loc), concrete(concrete)
+{
+  symbol.base_name=
+    "dynamic_object$"+std::to_string(loc->location_number)+suffix;
+  symbol.name="ssa::"+id2string(symbol.base_name);
+  symbol.is_lvalue=true;
+  symbol.type=type;
+  symbol.type.set("#dynamic", true);
+  symbol.mode=ID_C;
+}
+
+exprt dynamic_objectt::address_of(const typet &result_type) const
+{
+  typet object_type;
+  exprt object;
+
+  if(symbol.type.id()==ID_array)
+  {
+    object_type=symbol.type.subtype();
+    index_exprt index_expr(
+      symbol.symbol_expr(),
+      from_integer(0, index_type()),
+      symbol.type.subtype());
+    object=index_expr;
+    if(concrete)
+      to_index_expr(object).array().set("#concrete", true);
```

Comment:
Can't we also possibly get rid of this `#concrete` flag? I see it's used only in one place, `get_deallocated_precondition` on expressions obtained from `assignments::ssa_objects::globals`. From briefly looking, I believe we should be able to store the information about `#concrete` in the `globals` set (or in some other helper map) since it is constructed from a set of `ssa_objectt`, which we can use to obtain the symbol (and hence object name) and check in the object DB if it is a concrete object.